### PR TITLE
support force load config on classpath by "classpath:" prefix (like spring)

### PR DESCRIPTION
**Pull request by qxo**
and if start with '/' then skip URL logic

### CODE REVIEW

File: .ci/pitest.sh [hunk 1]
```diff
@@ -188,8 +188,9 @@ pitest-tree-walker)
 pitest-utils)
   mvn -e -P$1 clean test org.pitest:pitest-maven:mutationCoverage;
   declare -a ignoredItems=(
-  "CommonUtil.java.html:<td class='uncovered'><pre><span  class=''>                catch (final URISyntaxException ex) {</span></pre></td></tr>"
-  "CommonUtil.java.html:<td class='uncovered'><pre><span  class='survived'>                    throw new CheckstyleException(UNABLE_TO_FIND_EXCEPTION_PREFIX + filename, ex);</span></pre></td></tr>"
+  "CommonUtil.java.html:<td class='covered'><pre><span  class='survived'>            if (filename.charAt(0) != &#39;/&#39;) {</span></pre></td></tr>"
```
Comment (romani) [inline .ci/pitest.sh:191]: not acceptable to extend this technical debt, please cover it by UT.

File: src/main/java/com/puppycrawl/tools/checkstyle/utils/CommonUtil.java [hunk 2]
```diff
@@ -75,6 +75,9 @@
     /** Symbols with which multiple comment ends. */
     private static final String BLOCK_MULTIPLE_COMMENT_END = "*/";
 
+    /** Pseudo URL prefix for loading from the class path: "classpath:". */
+    private static final String CLASSPATH_URL_PREFIX = "classpath:";
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/utils/CommonUtil.java:79]: we need to add it to doc https://checkstyle.org/property_types.html#URI
website sources are in our repo, you can find them by grep.

File: src/test/java/com/puppycrawl/tools/checkstyle/utils/CommonUtilTest.java [hunk 3]
```diff
@@ -446,7 +447,29 @@ public void testGetUriByFilenameFindsAbsoluteResourceOnClasspath() throws Except
     }
 
     @Test
-    public void testGetUriByFilenameFindsRelativeResourceOnClasspath() throws Exception {
+    public void testGetUriByFilenameByClasspathPrefix() throws CheckstyleException {
+        final String filename =
+            "classpath:/" + getPackageLocation() + "/InputCommonUtilTest_empty_checks.xml";
```
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/utils/CommonUtilTest.java:452]: please update some (not all, as we need to be sure we keep compatibility) our configs to use this.
We have to use features we produce.

**Review by romani — CHANGES_REQUESTED** [review]
please spend time to explain in PR description all details on how it will work.

This PR have to pass several reviewers, so all should be clear transperent.

items to improve meanwhile:

File: src/main/java/com/puppycrawl/tools/checkstyle/utils/CommonUtil.java [hunk 4]
```diff
@@ -500,44 +503,54 @@ public static void close(Closeable closeable) {
...
-                    }
-                    if (configUrl == null) {
-                        throw new CheckstyleException(UNABLE_TO_FIND_EXCEPTION_PREFIX + filename);
-                    }
-                    uri = configUrl.toURI();
+                    final URL url = new URL(filename);
+                    uri = url.toURI();
+                }
+                catch (final URISyntaxException | MalformedURLException ignored) {
+                    // ignored
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/utils/CommonUtil.java:518]: No good, we need to rethrow all outside of Checkstyle

**Review by romani — CHANGES_REQUESTED** [review]
Item:

File: src/main/java/com/puppycrawl/tools/checkstyle/utils/CommonUtil.java (hunk 2 above)
Comment (qxo) [reply src/main/java/com/puppycrawl/tools/checkstyle/utils/CommonUtil.java:79]: > we need to add it to doc https://checkstyle.org/property_types.html#URI
> website sources are in our repo, you can find them by grep.

I'll try:)
please help me fix the doc(ie: grammar)

File: src/main/java/com/puppycrawl/tools/checkstyle/utils/CommonUtil.java (hunk 4 above)
Comment (qxo) [reply src/main/java/com/puppycrawl/tools/checkstyle/utils/CommonUtil.java:518]: it's history code. we should support windows path, ie: 'C:/xxx'

Or we change  precheck logic just on http ?
```java
            if (filename.startsWith("http://") || filename.startsWith("https://")) {
                try {
                    final URL url = new URL(filename);
                    uri = url.toURI();
                }
                catch (final URISyntaxException | MalformedURLException ignored) {
                    throw new CheckstyleException(UNABLE_TO_FIND_EXCEPTION_PREFIX + filename, ignored);
                }
            }
```

Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/utils/CommonUtil.java:518]: @qxo Don't mark thinks as resolved when you have a question back or the item isn't resolved from the initial review. Just say done when you believe you have resolved it and allow the reviewers to mark things resolved to confirm.

Comment (qxo) [reply src/main/java/com/puppycrawl/tools/checkstyle/utils/CommonUtil.java:518]: ok:)

File: .ci/pitest.sh (hunk 1 above)
Comment (qxo) [reply .ci/pitest.sh:191]: @romani I agree, But I can not fix this

Does any know how to create a valid URL from classpath resource  which it's toURI will throw a URISyntaxException ?

**Comment by nmancus1** [comment]
@qxo please share your thoughts at https://github.com/checkstyle/checkstyle/issues/9170

**Comment by romani** [comment]
implementation is moved to https://github.com/checkstyle/checkstyle/pull/9178

**Pull request closed without merging**